R/assert.R

```diff
@@ -46,3 +63,24 @@ vec_assert <- function(x, ptype = NULL, size = NULL) {
 
   invisible(x)
 }
+#' @rdname vec_assert
+#' @export
+vec_is <- function(x, ptype = NULL, size = NULL) {
+  if (!is_null(ptype)) {
+    ptype <- vec_type(ptype)
```

Comment:
I was imagining one more step:

```R
vec_is_size <- function(x, size) {
  if (is.null(size)) {
    return(TRUE
  }

  size <- vec_recycle(vec_cast(size, integer()), 1L)
  x_size <- vec_size(x)
  identical(x_size, size)
}
```

Then `vec_is()` is just `vec_is_ptype(x, ptype) && vec_is_size(x, size)` and same checks are used in `vec_assert()` to determine if an error needs to be thrown.

Reply:
This is what I did in https://github.com/r-lib/vctrs/pull/226/commits/05dd79df7274173fe7df4bd925bf240e122e4be4 but this also implies code duplication. I think the code duplication in the last commit makes a bit more sense?

Reply:
Oooh, I see now - you need the same calculations for the error messages.